### PR TITLE
DEV: Ability to scroll sidebar to specific element

### PR DESCRIPTION
**Pull request by lis2**
Ability to scroll to specific section in sidebar. Sometimes a sidebar might be too short, and we have to add height to the last element.

NOTE: code related to screen resolutions and scroll is a bit difficult to test

Related PR: https://github.com/discourse/discourse-chat/pull/1101

Demo:
https://user-images.githubusercontent.com/72780/181670236-a8109bbf-e158-4243-8296-4654cbfed079.mov

### CODE REVIEW

File: app/assets/javascripts/discourse/app/components/sidebar.js [hunk 1]
```diff
@@ -30,10 +31,43 @@ export default class Sidebar extends GlimmerComponent {
       this.args.toggleSidebar();
     }
   }
+  @bind
+  scrollToElement(destinationElement) {
+    const topPadding = 10;
+    const sidebarContainerElement =
+      document.getElementsByClassName("sidebar-container")[0];
+    const sidebarSectionsElement =
+      document.getElementsByClassName("sidebar-sections")[0];
```
Comment (tgxworld) [inline app/assets/javascripts/discourse/app/components/sidebar.js:40]:
```suggestion
    const sidebarContainerElement =
      document.querySelector(".sidebar-container");

    const sidebarSectionsElement =
      document.querySelector(".sidebar-sections");
```

File: app/assets/javascripts/discourse/app/components/sidebar.js [hunk 2]
```diff
@@ -30,10 +31,43 @@ export default class Sidebar extends GlimmerComponent {
       this.args.toggleSidebar();
     }
   }
+  @bind
+  scrollToElement(destinationElement) {
+    const topPadding = 10;
```
Comment (tgxworld) [inline app/assets/javascripts/discourse/app/components/sidebar.js:36]: Is the value here dependent on somewhere else?

Comment (lis2) [reply app/assets/javascripts/discourse/app/components/sidebar.js:36]: no, this was my way of "no magic number" rule. So kind of config value at the top, but not used anywhere else, so don't need exported constant

File: app/assets/javascripts/discourse/app/components/sidebar.js [hunk 3]
```diff
@@ -8,6 +8,7 @@ export default class Sidebar extends GlimmerComponent {
     if (this.site.mobileView) {
       document.addEventListener("click", this.collapseSidebar);
     }
+    this.appEvents.on("sidebar:scroll-to-element", this.scrollToElement);
```
Comment (tgxworld) [inline app/assets/javascripts/discourse/app/components/sidebar.js:11]:
```suggestion
    // This appEvent handler is experimental and should not be relied on as an extension point yet.
    this.appEvents.on("sidebar:scroll-to-element", this, this.#scrollToElement);
```

File: app/assets/javascripts/discourse/app/components/sidebar.js [hunk 4]
```diff
@@ -30,10 +31,58 @@ export default class Sidebar extends GlimmerComponent {
       this.args.toggleSidebar();
     }
   }
+  @bind
+  scrollToElement(destinationElement) {
```
Comment (tgxworld) [inline app/assets/javascripts/discourse/app/components/sidebar.js:35]:
```suggestion
  #scrollToElement(destinationElement) {
```

File: app/assets/javascripts/discourse/app/components/sidebar.js [hunk 5]
```diff
@@ -30,10 +32,58 @@ export default class Sidebar extends GlimmerComponent {
...
+    const distanceFromTop =
+      document.getElementsByClassName(destinationElement)[0].offsetTop -
+      topPadding;
+
+    this.setMissingHeightForScroll(sidebarContainerElement, distanceFromTop);
+
+    sidebarContainerElement.scrollTop = distanceFromTop;
+  }
+
+  setMissingHeightForScroll(sidebarContainerElement, distanceFromTop) {
```
Comment (tgxworld) [inline app/assets/javascripts/discourse/app/components/sidebar.js:49]:
```suggestion
  #setMissingHeightForScroll(sidebarContainerElement, distanceFromTop) {
```

File: app/assets/javascripts/discourse/app/components/sidebar.js [hunk 6]
```diff
@@ -30,10 +32,58 @@ export default class Sidebar extends GlimmerComponent {
...
+  @bind
+  scrollToElement(destinationElement) {
+    const topPadding = 10;
+    const sidebarContainerElement =
+      document.querySelector(".sidebar-container");
+    const distanceFromTop =
+      document.getElementsByClassName(destinationElement)[0].offsetTop -
+      topPadding;
+
+    this.setMissingHeightForScroll(sidebarContainerElement, distanceFromTop);
```
Comment (tgxworld) [inline app/assets/javascripts/discourse/app/components/sidebar.js:44]:
```suggestion
    this.#setMissingHeightForScroll(sidebarContainerElement, distanceFromTop);
```

File: app/assets/javascripts/discourse/app/components/sidebar.js [hunk 7]
```diff
@@ -30,10 +32,58 @@ export default class Sidebar extends GlimmerComponent {
...
+
+    lastSectionElement.style.paddingBottom =
+      missingHeight > 0 ? `${missingHeight}px` : null;
+  }
 
   willDestroy() {
     if (this.site.mobileView) {
       document.removeEventListener("click", this.collapseSidebar);
     }
+    this.appEvents.off("sidebar:scroll-to-element", this.scrollToElement);
```
Comment (tgxworld) [inline app/assets/javascripts/discourse/app/components/sidebar.js:87]:
```suggestion
    this.appEvents.off("sidebar:scroll-to-element", this, this.#scrollToElement);
```

File: app/assets/javascripts/discourse/app/components/sidebar.js [hunk 8]
```diff
@@ -30,10 +32,57 @@ export default class Sidebar extends GlimmerComponent {
...
+
+    const missingHeight =
+      sidebarContainerElement.clientHeight -
+      headerOffset +
+      lastSectionBottomPadding -
+      (allSectionsHeight - distanceFromTop);
+
+    lastSectionElement.style.paddingBottom =
+      missingHeight > 0 ? `${missingHeight}px` : null;
+  }
```
Comment (tgxworld) [inline app/assets/javascripts/discourse/app/components/sidebar.js:80]:
```suggestion
  #setMissingHeightForScroll(sidebarContainerElement, distanceFromTop) {
    const allSections = document.getElementsByClassName(
      "sidebar-section-wrapper"
    );

    const lastSectionElement = allSections[allSections.length - 1];

    const lastSectionBottomPadding = parseInt(
      lastSectionElement.style.paddingBottom?.replace("px", "") || 0,
      10
    );

    const headerOffset = parseInt(
      document.documentElement.style.getPropertyValue("--header-offset"),
      10
    );

    let allSectionsHeight = 0;

    for (const section of allSections) {
      allSectionsHeight +=
        section.clientHeight +
        parseInt(
          window.getComputedStyle(section).marginBottom.replace("px", ""),
          10
        );
    }

    const missingHeight =
      sidebarContainerElement.clientHeight -
      headerOffset +
      lastSectionBottomPadding -
      (allSectionsHeight - distanceFromTop);

    lastSectionElement.style.paddingBottom =
      missingHeight > 0 ? `${missingHeight}px` : null;
  }
```

File: app/assets/javascripts/discourse/app/components/sidebar.js [hunk 9]
```diff
@@ -30,10 +32,61 @@ export default class Sidebar extends GlimmerComponent {
...
+    const sidebarContainerElement =
+      document.querySelector(".sidebar-container");
+    const distanceFromTop =
+      document.getElementsByClassName(destinationElement)[0].offsetTop -
+      topPadding;
+
+    this.#setMissingHeightForScroll(sidebarContainerElement, distanceFromTop);
+
+    sidebarContainerElement.scrollTop = distanceFromTop;
+  }
```
Comment (tgxworld) [inline app/assets/javascripts/discourse/app/components/sidebar.js:46]:
```suggestion
  #scrollToElement(destinationElement) {
    const topPadding = 10;

    const sidebarContainerElement =
      document.querySelector(".sidebar-container");
      
    const distanceFromTop =
      document.getElementsByClassName(destinationElement)[0].offsetTop -
      topPadding;

    this.#setMissingHeightForScroll(sidebarContainerElement, distanceFromTop);

    sidebarContainerElement.scrollTop = distanceFromTop;
  }
``````suggestion
  #scrollToElement(destinationElement) {
    const topPadding = 10;

    const sidebarContainerElement =
      document.querySelector(".sidebar-container");
      
    const distanceFromTop =
      document.getElementsByClassName(destinationElement)[0].offsetTop -
      topPadding;

    this.#setMissingHeightForScroll(sidebarContainerElement, distanceFromTop);

    sidebarContainerElement.scrollTop = distanceFromTop;
  }
```

**Review by tgxworld — APPROVED** [review]

File: app/assets/javascripts/discourse/app/components/sidebar.js [hunk 10]
```diff
@@ -30,10 +32,78 @@ export default class Sidebar extends GlimmerComponent {
...
+      document.querySelector(".sidebar-container");
+      
+    const distanceFromTop =
+      document.getElementsByClassName(destinationElement)[0].offsetTop -
+      topPadding;
+
+    this.#setMissingHeightForScroll(sidebarContainerElement, distanceFromTop);
+
+    sidebarContainerElement.scrollTop = distanceFromTop;
+  }
```
Comment (tgxworld) [inline app/assets/javascripts/discourse/app/components/sidebar.js:63]:
```suggestion
  #scrollToElement(destinationElement) {
    const topPadding = 10;

    const sidebarContainerElement =
      document.querySelector(".sidebar-container");
      
    const distanceFromTop =
      document.getElementsByClassName(destinationElement)[0].offsetTop -
      topPadding;

    this.#setMissingHeightForScroll(sidebarContainerElement, distanceFromTop);

    sidebarContainerElement.scrollTop = distanceFromTop;
  }
```

**Comment by davidtaylorhq** [comment]
Is there a reason that we can't use `Element.scrollIntoView()` here. That way, we wouldn't need to manually calculate the height? https://developer.mozilla.org/en-US/docs/Web/API/Element/scrollIntoView

(we already ship a polyfill to enable the 'smooth' behaviour on older versions of Safari)

**Comment by lis2** [comment]
> Is there a reason that we can't use `Element.scrollIntoView()` here.

Nice idea, I didn't know about that function. However, the problem still exists when the sidebar is too short.

From documentation: "The element may not be scrolled completely to the top or bottom depending on the layout of other elements."

In that case, I still have to calculate missing height and add it to the last element

image

File: app/assets/javascripts/discourse/app/components/sidebar.js [hunk 11]
```diff
@@ -30,10 +32,67 @@ export default class Sidebar extends GlimmerComponent {
       this.args.toggleSidebar();
     }
   }
+  #scrollToElement(destinationElement) {
```
Comment (tgxworld) [inline app/assets/javascripts/discourse/app/components/sidebar.js:35]:
```suggestion

  #scrollToElement(destinationElement) {
```

File: app/assets/javascripts/discourse/app/components/sidebar.js [hunk 12]
```diff
@@ -31,9 +33,67 @@ export default class Sidebar extends GlimmerComponent {
...
+
+    lastSectionElement.style.paddingBottom =
+      missingHeight > 0 ? `${missingHeight}px` : null;
+  }
+
   willDestroy() {
     if (this.site.mobileView) {
       document.removeEventListener("click", this.collapseSidebar);
     }
+    this.appEvents.off(
```
Comment (tgxworld) [inline app/assets/javascripts/discourse/app/components/sidebar.js:93]:
```suggestion

    this.appEvents.off(
```

**Pull request merged**